File: app/src/main/java/net/tlalka/android/fiszki/view/activities/LessonActivity.java
package net.tlalka.android.fiszki.view.activities;

import java.util.List;
import javax.inject.Inject;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.OnClick;
import net.tlalka.android.fiszki.R;
import net.tlalka.android.fiszki.domain.controllers.LessonController;
import net.tlalka.android.fiszki.domain.utils.ValidUtils;
import net.tlalka.android.fiszki.model.dto.LessonDto;
import net.tlalka.android.fiszki.model.entities.Word;
import net.tlalka.android.fiszki.model.types.LanguageType;
import net.tlalka.android.fiszki.view.fragments.LanguageDialogFragment;
import net.tlalka.android.fiszki.view.navigations.Navigator;

public class LessonActivity extends BasePageActivity implements LanguageDialogFragment.DialogListener {

    @BindView(R.id.lesson_topic)
    protected TextView lessonTopic;

    @BindView(R.id.lesson_progress)
    protected TextView lessonProgress;

    @BindView(R.id.lesson_show_word)
    protected Button lessonShowWord;

    @BindView(R.id.lesson_check_word)
    protected Button lessonCheckWord;

    @Inject
    protected LessonController lessonController;

    @Inject
    protected Navigator navigator;

    @Inject
    protected LessonDto lessonDto;

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        super.setContentView(R.layout.lesson_activity);
        super.getActivityComponent().inject(this);

        this.runActivity();
    }

    private void runActivity() {
        int lessonIndex = this.lessonDto.getLessonIndex();
        String lessonName = this.lessonDto.getLessonName();

        this.lessonTopic.setText(getString(R.string.lesson_topic, lessonIndex, lessonName));
        this.generateView();
    }

    private void generateView() {
        if (this.lessonController.hasNextWord()) {
            this.lessonProgress.setText(lessonController.getLessonStatus());
            this.lessonShowWord.setText(lessonController.getNextWord());
            this.lessonCheckWord.setText(getText(R.string.lesson_check_word));
        } else {
            this.showLessonSummary();
        }
    }

    private void showLessonSummary() {
        this.lessonController.updateLessonDto(lessonDto);
        this.lessonController.updateLessonProgress();
        this.navigator.openLessonScoreActivity(this, lessonDto);
        this.navigator.finish(this);
    }

    @OnClick(R.id.lesson_check_word)
    public void onCheckWordClick(View view) {
        Word word = this.lessonController.getTranslateWord();

        if (ValidUtils.isNotNull(word)) {
            this.lessonCheckWord.setText(word.getValue());
        } else {
            List<LanguageType> languages = lessonController.getLanguages();
            LanguageDialogFragment dialog = LanguageDialogFragment.getInstance(languages);
            dialog.show(getFragmentManager(), LanguageDialogFragment.class.getName());
        }
    }

    @Override
    public void onLanguageSelected(LanguageType languageType) {
        this.lessonController.setTranslation(languageType);
        this.onCheckWordClick(lessonCheckWord);
    }

    @OnClick(R.id.button_correct)
    public void onCorrectClick(View view) {
        this.lessonController.correctAnswer();
        this.generateView();
    }

    @OnClick(R.id.button_incorrect)
    public void onIncorrectClick(View view) {
        this.lessonController.incorrectAnswer();
        this.generateView();
    }
}

File: app/src/main/java/net/tlalka/android/fiszki/view/activities/LessonActivity.kt
package net.tlalka.android.fiszki.view.activities

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import butterknife.BindView
import butterknife.OnClick
import net.tlalka.android.fiszki.R
import net.tlalka.android.fiszki.domain.controllers.LessonController
import net.tlalka.android.fiszki.domain.utils.ValidUtils
import net.tlalka.android.fiszki.model.dto.LessonDto
import net.tlalka.android.fiszki.model.types.LanguageType
import net.tlalka.android.fiszki.view.fragments.LanguageDialogFragment
import net.tlalka.android.fiszki.view.navigations.Navigator
import javax.inject.Inject

class LessonActivity : BasePageActivity(), LanguageDialogFragment.DialogListener {

    @BindView(R.id.lesson_topic)
    lateinit var lessonTopic: TextView

    @BindView(R.id.lesson_progress)
    lateinit var lessonProgress: TextView

    @BindView(R.id.lesson_show_word)
    lateinit var lessonShowWord: Button

    @BindView(R.id.lesson_check_word)
    lateinit var lessonCheckWord: Button

    @Inject
    lateinit var lessonController: LessonController

    @Inject
    lateinit var navigator: Navigator

    @Inject
    lateinit var lessonDto: LessonDto

    public override fun onCreate(bundle: Bundle?) {
        super.onCreate(bundle)
        super.setContentView(R.layout.lesson_activity)
        super.activityComponent.inject(this)

        runActivity()
    }

    private fun runActivity() {
        lessonTopic.text = getString(R.string.lesson_topic, lessonDto.lessonIndex, lessonDto.lessonName)
        generateView()
    }

    private fun generateView() {
        if (lessonController.hasNextWord()) {
            lessonProgress.text = lessonController.getLessonStatus()
            lessonShowWord.text = lessonController.getNextWord()
            lessonCheckWord.text = getText(R.string.lesson_check_word)
        } else {
            showLessonSummary()
        }
    }

    private fun showLessonSummary() {
        lessonController.updateLessonDto(lessonDto)
        lessonController.updateLessonProgress()
        navigator.openLessonScoreActivity(this, lessonDto)
        navigator.finish(this)
    }

    @OnClick(R.id.lesson_check_word)
    fun onCheckWordClick(@Suppress("UNUSED_PARAMETER") view: View?) {
        val word = this.lessonController.getTranslateWord()

        if (ValidUtils.isNotNull(word)) {
            lessonCheckWord.text = word.value
        } else {
            LanguageDialogFragment
                    .getInstance(lessonController.getLanguages())
                    .show(fragmentManager, LanguageDialogFragment::class.java.name)
        }
    }

    override fun onLanguageSelected(languageType: LanguageType) {
        lessonController.setTranslation(languageType)
        onCheckWordClick(lessonCheckWord)
    }

    @OnClick(R.id.button_correct)
    fun onCorrectClick(view: View) {
        lessonController.correctAnswer()
        generateView()
    }

    @OnClick(R.id.button_incorrect)
    fun onIncorrectClick(view: View) {
        lessonController.incorrectAnswer()
        generateView()
    }
}
